components/settings/settings.cpp

```diff
@@ -117,6 +120,230 @@ class SettingsFileParser
+
+            // Settings not in the written map are definitely invalid.  Currently, this can only
+            // happen if the player edited the file while playing, because loadSettingsFile()
+            // will accept anything and pass it along in the map, but in the future, we might
+            // want to handle invalid settings more gracefully here.
+            if (finder == written.end()) {
+                ostream << "# invalid setting: " << line << std::endl;
+                changed = true;
+                continue;
+            }
+
+            // Write the current value of the setting to the file.  The key must exist in the
+            // settings map because of how written was initialized and finder != end().
+            ostream << setting << " = " << settings[key] << std::endl;
+            // Mark that setting as written.
+            finder->second = true;
+            // Did we really change it?
+            if (value != settings[key]) {
+                std::cout << "Changed setting: [" << currentCategory << "] "
+                          << setting << " = " << settings[key] << std::endl;
+                changed = true;
+            }
+            // No need to write the current line, because we just emitted a replacement.
+
+            // Curiously, it appears that comments at the ends of lines with settings are not
+            // allowed, and the comment becomes part of the value.  Was that intended?
+        }
+
+        // We're done with the input stream file.
+        istream.close();
+
+        // Ensure that all options in the current category have been written.  We must complete
+        // the current category at the end of the file before moving on to any new categories.
+        for (CategorySettingStatusMap::iterator mit = written.begin(); mit != written.end(); ++mit) {
+            bool missed = false;
+            if (mit->second == false && mit->first.first == currentCategory) {
+                std::cout << "Added new setting: [" << mit->first.first << "] "
+                          << mit->first.second << " = " << settings[mit->first] << std::endl;
+                // With some further code changes, this comment could be more descriptive.
+                ostream << std::endl << "# Automatically added setting." << std::endl;
+                ostream << mit->first.second << " = " << settings[mit->first] << std::endl;
+                mit->second = true;
+                missed = true;
+                changed = true;
+            }
+            // Ensure that there's a newline before the next section if we added settings.
+            if (missed) ostream << std::endl;
+        }
+
+        // This logic triggers when the input file was effectively empty.  It could be extended
+        // to doing something more intelligent instead, like make a copy of the default settings
+        // file (complete with comments) before continuing.  Other code prevents OpenMW from
```

Comment:
> Was not aware of that. Definitely not intended. One more reason to nuke the launcher settings code and make it use the existing settings component.

Agreed.  Does that mean I'm on the right track to write a wrapper for the working settings code to make it use QStrings and QMaps for the Launcher?   If so I'll start working on that.  My plan was to convert the one data structure to the other and maintain the existing API with the exception of passing an QIOstream.   I intend to pass file name instead and wrap the call in a try block to generate the nice Qt warning dialog.

> I have to assume you pulled the latest code, but didn't actually compile it.

That's likely.  But the settings-default.cfg would have been the old one as well, correct?  Perhaps there was another factor in play as well (like using or not using the launcher).

> As far as I could tell, your branch does behave correct w.r.t. not copying the complete default settings into user settings (barring the launcher bug). I just mentioned the issue because the comment on this line suggested that as a future possibility:

Yes, I thought that only updating existing options was the right approach as well.  The comment  was in part because I thought that's how you wanted it to work.   I'm clear on what the desired behavior is now.   I still have one question though:

If a user changes a setting to a non-default value, and then changes it back to a default one, do we want to remove it from the user configuration?   Is there a data structure in the code that says what the default value is?  If so, I don't know where it is.  I supposed I could read the default-settings.cfg and make a different Settings map to compare to...  My current code uses the approach that whatever is in the user configuration is what should be written back out to the user configuration.  Conceptually it is possible to detect that the setting has reverted to the default value, and then remove both the setting and the comments since the last setting from the user configuration.   That would be very clear and consistent with the overall approach in my opinion.  Is this what you want, or is the plan that once user configured, always user configured?